pkg/gofr/datasource/opentsdb/annotation.go

```diff
@@ -0,0 +1,334 @@
+// via the /api/annotation/bulk endpoint.
+type BulkAnnotatResponse struct {
+	// StatusCode holds the HTTP status code of the bulk annotation request.
+	StatusCode int
+
+	// UpdateAnnotations holds the list of annotations involved in a bulk update.
+	UpdateAnnotations []Annotation `json:"InvolvedAnnotations,omitempty"`
+
+	// ErrorInfo contains details about any errors that occurred during the bulk operation.
+	ErrorInfo map[string]interface{} `json:"error,omitempty"`
+
+	// Tsuids holds the list of TSUIDs for annotations that should be deleted.
+	// If empty or nil, the global flag is used.
+	Tsuids []string `json:"tsuids,omitempty"`
+
+	// StartTime is the Unix epoch timestamp for the start of the deletion request.
+	StartTime int64 `json:"startTime,omitempty"`
+
+	// EndTime is the optional Unix epoch timestamp for when the events ended.
+	EndTime int64 `json:"endTime,omitempty"`
+
+	// Global indicates whether global annotations should be deleted for the given time range.
+	Global bool `json:"global,omitempty"`
+
+	// TotalDeleted holds the total number of annotations successfully deleted in the bulk operation.
+	TotalDeleted int64 `json:"totalDeleted,omitempty"`
+
+	logger Logger
+	tracer trace.Tracer
+	ctx    context.Context
+}
+
+// BulkAnnoDeleteInfo holds the parameters for a bulk annotation delete operation.
+type BulkAnnoDeleteInfo struct {
+	// Tsuids holds the list of TSUIDs for annotations that should be deleted.
+	// If empty or nil, the global flag is used.
+	Tsuids []string `json:"tsuids,omitempty"`
+
+	// StartTime is the Unix epoch timestamp for the start of the deletion request.
+	StartTime int64 `json:"startTime,omitempty"`
+
+	// EndTime is the optional Unix epoch timestamp for when the events ended.
+	EndTime int64 `json:"endTime,omitempty"`
+
+	// Global indicates whether global annotations should be deleted for the given time range.
+	Global bool `json:"global,omitempty"`
+}
+
+// BulkDeleteResp contains the results of a bulk annotation delete operation.
+type BulkDeleteResp struct {
+	BulkAnnoDeleteInfo
+}
+
+func (bulkAnnotResp *BulkAnnotatResponse) SetStatus(code int) {
+	setStatus(bulkAnnotResp.ctx, bulkAnnotResp, code, "SetStatus-BulkAnnotation", bulkAnnotResp.logger)
+}
+
+func (bulkAnnotResp *BulkAnnotatResponse) setStatusCode(code int) {
+	bulkAnnotResp.StatusCode = code
+}
+
+func (bulkAnnotResp *BulkAnnotatResponse) GetCustomParser() func(respCnt []byte) error {
+	return getCustomParser(bulkAnnotResp.ctx, bulkAnnotResp, "GetCustomParser-BulkAnnotation", bulkAnnotResp.logger,
+		func(resp []byte) error {
+			originContents := string(resp)
+
+			var resultBytes []byte
+
+			if strings.Contains(originContents, "error") || strings.Contains(originContents, "totalDeleted") {
+				resultBytes = resp
+			} else if strings.Contains(originContents, "startTime") {
+				resultBytes = []byte(fmt.Sprintf("{%s:%s}", `"InvolvedAnnotations"`, originContents))
+			} else {
```

Comment:
> I would decode in a struct with the field you need, then apply the logic of checking the fields

Here it gives error when error or totalDeleted is returned, in case of DeleteAnnotation ....while a struct starting with startTime or an error is returned in case of UpdateAnnotation. I tried generalizing the custom-parser for both of the operations.

Reply:
First, please bare with me while I have zero knowledge with OpenTSDB

But, imagine the code returns a valid reply that contains the "error" string, not as a JSON node, but somewhere in the reply.

Your code will fail.

Reply:
@coolwednesday I feel the same, would test with a few more cases and check.